File: services/graphql-service/app/application/use_cases/user/update_user_use_case.py
from uuid import UUID

from app.domain.entities.user import User
from app.application.interfaces.user_repository_interface import IUserRepository
from app.application.dto.update_user_dto import UpdateUserDTO

class UpdateUserUseCase:

    def __init__(
        self,
        repository: IUserRepository
    ):
        self.repository = repository

    async def execute(
        self,
        dto: UpdateUserDTO
    ) -> User | None:
        
        return await self.repository.update(dto)

File: services/grpc-service/src/GrpcService.Application/UseCases/UpdateUser/UpdateUserUseCase.cs
using GrpcService.Application.Dtos;
using GrpcService.Application.Interfaces;
using GrpcService.Application.Mappers;
using GrpcService.Domain.Exceptions;

namespace GrpcService.Application.UseCases.UpdateUser;

public sealed class UpdateUserUseCase(IUserRepository repository)
{
    public async Task<UserDto> ExecuteAsync(
        UpdateUserDto dto, 
        CancellationToken cancellationToken = default
    )
    {
        var user = await repository.UpdateAsync(
            dto.Id, dto.Name, dto.Email, dto.Age, cancellationToken);

        if (user is null)
            throw new UserNotFoundException(dto.Id);

        return user.ToDto();
    }
}